src/ufo.rs

```diff
@@ -194,6 +193,39 @@ impl Ufo {
                 .collect();
             let layers = layers?;
 
+            match meta.format_version {
+                FormatVersion::V1 | FormatVersion::V2 => {
+                    // Upconvert UFO v1 or v2 kerning if necessary. To upconvert, we need at least
+                    // a groups.plist file...
+                    if let Some(groups_to_upconvert) = &groups {
+                        // ... while a kerning.plist is optional. If there isn't one, pass an empty
+                        // map to the converter to keep the code simple.
+                        let kerning_to_upconvert;
+                        let kerning_empty = BTreeMap::new();
+                        if let Some(kerning_some) = &kerning {
+                            kerning_to_upconvert = kerning_some
+                        } else {
+                            kerning_to_upconvert = &kerning_empty
+                        };
+
+                        let (groups_new, kerning_new) = upconvert_kerning(
+                            &groups_to_upconvert,
+                            &kerning_to_upconvert,
+                            &glyph_names,
+                        );
+                        match validate_groups(&groups_new) {
+                            Err(Error::Groups(e)) => Err(Error::GroupsUpconversionError(e))?,
+                            _ => {}
+                        }
```

Comment:
Hmm, this is a bit funny.

First: if `validate_groups` only ever returns a `GroupsError`, it should return that directly as its error type, instead of wrapping it in `Error`; the caller can then convert to `Error` as needed.

The normal way to do this is to implement `From` for your inner error type to your outer error type; then the caller can just use `?` and it will convert the error types as needed.

That no longer makes sense here, since there are two different outer error types that use the same inner type, and so you'll have to be explicit. In any case, what I *think* this should look like is,

```rust
validate_groups(&groups_new).map_err(|e| Error::GroupsUpConversionError(e))?;
```

If you wanted to get _extra_ fancy, you could have some simple methods on`GroupsError` that would convert it to the different variants of the outer error; then you could use those in your `map_err`, for instance:

```rust
impl GroupsError {
    fn into_upconversion_error(self) -> Error {
        Error::GroupsUpConversionError(self)
    }
}

// and then
validate_groups(&groups_new).map_err(GroupsError::into_upconversion_error)?;
```